src/operators_lazyproduct.jl

```diff
@@ -10,23 +10,34 @@ The factors of the product are stored in the `operators` field. Additionally a
 complex factor is stored in the `factor` field which allows for fast
 multiplication with numbers.
 """
-mutable struct LazyProduct{BL,BR,F,T} <: AbstractOperator{BL,BR}
+
+mutable struct LazyProduct{BL,BR,F,T,KTL,BTR} <: AbstractOperator{BL,BR}
     basis_l::BL
     basis_r::BR
     factor::F
     operators::T
-    function LazyProduct{BL,BR,F,T}(operators::T, factor::F=1) where {BL,BR,F,T}
+    ket_l::KTL
+    bra_r::BTR
+    function LazyProduct{BL,BR,F,T,KTL,BTR}(operators::T, factor::F=1) where {BL,BR,F,T,KTL,BTR}
         for i = 2:length(operators)
             check_multiplicable(operators[i-1], operators[i])
         end
-        new(operators[1].basis_l, operators[end].basis_r, factor, operators)
+        ket_l=[Ket(operator.basis_l) for operator in operators]
+        bra_r=[Bra(operator.basis_r) for operator in operators]
+        new(operators[1].basis_l, operators[end].basis_r, factor, operators,ket_l,bra_r)
     end
 end
 function LazyProduct(operators::T, factor::F=1) where {T,F}
     BL = typeof(operators[1].basis_l)
     BR = typeof(operators[end].basis_r)
-    LazyProduct{BL,BR,F,T}(operators, factor)
+    KTL = typeof([Ket(operator.basis_l) for operator in operators])
+    BTR = typeof([Bra(operator.basis_r) for operator in operators])
```

Comment:
This will cause the entire list of `Ket`s and `Bra`s to be calculated twice when constructing a `LazyProduct`. Can you do the type inference and construction of the cache in one go?

Reply:
OK, it can be quickly resovled.